Draft changes for packaging/deployment/lifecycle. (#540)
eclipse-ee4j/cdi: spec/src/main/asciidoc/core/packagingdeployment_lite.asciidoc

== Packaging and deployment

> When an application is started, the container must perform _bean discovery_, detect definition errors and deployment problems and raise events that allow build compatible extensions to integrate with the container initialization lifecycle.

Author: Good point. Any suggestions on how to reformulate it?
Maybe just: `... detect definition errors and deployment problems and allow build compatible extensions to integrate with the container initialization lifecycle.`?

> When an application is started, the container must perform _bean discovery_, detect definition errors and deployment problems and raise events that allow build compatible extensions to integrate with the container initialization lifecycle.

Reviewer: ```suggestion
Before an application is started, or during application startup, the container must perform _bean discovery_, execute build compatible extensions., and detect definition errors and deployment problems.
```

Or maybe we should say "during application deployment", using the term "deployment" in a very abstract sense.

Bean discovery is the process of determining:

* Bean archives within application, and any beans contained within them
* Which alternatives and interceptors are _enabled_
* The _ordering_ of enabled interceptors

Additional beans may be registered programmatically via <<spi_lite>>.

[[bean_archive_lite]]

=== Bean archives

Bean classes of enabled beans must be deployed in _bean archives_.

A bean archive has a _bean discovery mode_ of either `annotated` or `none`.
This is governed by presence of `beans.xml` file which can be either empty or it can declare the `bean-discovery-mode` attribute.
Default value for this attribute is `annotated`.

An archive which:

* contains a `beans.xml` file with the `bean-discovery-mode` of `none`, or,
* contains a portable or build compatible extension and no `beans.xml` file

> is not a bean archive.

Author: This wording is the same as in previous spec versions and same contradiction exists in its Full variant.
I agree that it sounds weird but you technically have to consider an archive to be a bean archive due to presence of `beans.xml` which you then parse only to find out it actually isn't a bean archive :-D

I mean if we instead said that:
> An archive has a _bean discovery mode_ of either `annotated` or `none`.

That would imply that all archives have to have some discovery mode which isn't true either, right?

An _implicit bean archive_ is:

* an archive which contains a `beans.xml` file that is empty, or,
* an archive which contains a `beanx.xml` file that has `bean-discovery-mode` attribute set to `annotated`

When determining which archives are bean archives, the container must consider:

* Library jars
* Directories in the JVM classpath

> Non Jakarta EE containers may or may not provide support for war, EJB jar or rar bean archives.

Reviewer: I'd just get rid of this completely.

IMHO, CDI Lite implementations should be required to just document how they search for bean archives, but we can't specify that exactly.

Something like:

```
When determining which archives are bean archives, the container must consider all JAR archives available through implementation-specific means.
Implementations are required to document how the candidate JAR archives are found.
```

Author: I like the wording! Do you think we should do something similar in the Full variant or do we keep it as is?

The `beans.xml` file must be named:

* `META-INF/beans.xml`.

> The container searches for beans in all bean archives in the application classpath.

Reviewer: Classpath is again a runtime concept. Following my suggestion above, I'd just say:

```
The container searches for beans in all bean archives discovered as described above.
```

If a bean class is deployed in two different bean archives, non-portable behavior results.
Portable applications must deploy each bean class in no more than one bean archive.

Implicit bean archives are likely to contain classes which are not deployed as beans.

An extension may be deployed in any archive, including those that are not bean archives.

[[initialization_lite]]

=== Application initialization lifecycle

> // TODO related to https://github.com/eclipse-ee4j/cdi/issues/482

Reviewer: Also related to https://github.com/eclipse-ee4j/cdi/issues/496
